sel/src/Sel/PublicKey/Signature.hs

```diff
@@ -163,8 +163,8 @@ generateKeyPair = do
 -- | Sign a message.
 --
 -- @since 0.0.1.0
-signMessage :: StrictByteString -> SecretKey -> IO SignedMessage
-signMessage message (SecretKey skFPtr) =
+signMessage :: StrictByteString -> SecretKey -> SignedMessage
+signMessage message (SecretKey skFPtr) = unsafeDupablePerformIO $
```

Comment:
This one is a bit questionable because of [this snippet in `libsodium`](https://github.com/jedisct1/libsodium/blob/3e12b517ee42eb4993d19611df325a01104709cc/src/libsodium/crypto_sign/ed25519/ref10/sign.c#L63-L67):

```c
#ifdef ED25519_NONDETERMINISTIC
    _crypto_sign_ed25519_synthetic_r_hv(&hs, nonce /* tmp */, az);
#else
    crypto_hash_sha512_update(&hs, az + 32, 32);
#endif
```

As mentioned in the[ `libsodium` docs](https://doc.libsodium.org/public-key_cryptography/public-key_signatures#notes), non-deterministic `Ed25519` signatures are non-standard, so it's doubtful that most people are building the library with this macro defined.

However, it's totally possible that it is defined. So, unfortunately, maybe it isn't safe to define this as pure :disappointed:

Reply:
Hmm yeah, maybe let's keep that one with `IO`, and we'll indicate that in the absence of `ED25519_NONDETERMINISTIC` it's ± safe to use `unsafeDupablePerformIO`